src/core/player.cpp

```diff
@@ -213,7 +215,19 @@ void Player::NextItem(Engine::TrackChangeFlags change) {
   // Manual track changes override "Repeat track"
   const bool ignore_repeat_track = change & Engine::Manual;
 
-  int i = active_playlist->next_row(ignore_repeat_track);
+  int i = active_playlist->current_row();
+  if (next_album && i != -1) {
+      QString album = active_playlist->current_item_metadata().album();
+
+      i = active_playlist->next_row(ignore_repeat_track);
+      while (active_playlist->item_at(i)->Metadata().album() == album) {
+          active_playlist->set_current_row(i,true);
+          i = active_playlist->next_row(ignore_repeat_track);
```

Comment:
If you have a single album in your playlist, and you have the playlist to repeat, will this loop forever?

Reply:
Good test case - I tried it and it causes clementine to crash. I'll have to think of a fix for this.

Reply:
Maybe Playlist could have a find_next_row (or FindNextRow.. there's a lot of inconsistency in method naming in there) that takes a functor that returns true if a match is found. Then you could just do something like:

```i = active_playlist->find_next_row([](PlaylistItemPtr& item) -> bool { return item->Metadata().album() != album});```

The method would return -1 if it looped through without finding a match.

Reply:
I found a shortcut to achieve the same behavior, "next_row()" return -1 when it tries to access the row beyond the last one on the list. I made that part of the while condition and tested it out. See my next commit.

Reply:
But I think it only returns -1 if you're not in playllist repeat mode. Otherwise, it loops (default case)
```
  if (next_virtual_index >= virtual_items_.count()) {
    // We've gone off the end of the playlist.

    switch (playlist_sequence_->repeat_mode()) {
      case PlaylistSequence::Repeat_Off:
      case PlaylistSequence::Repeat_Intro:
        return -1;
      case PlaylistSequence::Repeat_Track:
        next_virtual_index = current_virtual_index_;
        break;

      default:
        next_virtual_index = NextVirtualIndex(-1, ignore_repeat_track);
        break;
    }
```

Also note that NextVirtualIndex() check or album repeat mode.

Reply:
Thanks Jim. I'm going to rework the code per your original suggestion and create a new function to search for the next track rather than fiddling with the existing function.